cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/MetadataClient.java

```diff
@@ -1,126 +1,163 @@
         LOG.warn("Failed to update metadata in {}ms. Using old metadata with version {} and last successful update {}.",
-                 timeout, _metadata.version(), _metadata.lastSuccessfulUpdate());
+            timeoutMs, _version, _lastSuccessfulUpdateMs);
       }
     }
-    return new ClusterAndGeneration(_metadata.fetch(), _metadataGeneration.get());
+    return new ClusterAndGeneration(_cluster, _metadataGeneration.get());
   }
 
-  /**
-   * Close the metadata client. Synchronized to avoid interrupting the network client during a poll.
-   */
-  public synchronized void close() {
-    _networkClient.close();
+  private Cluster doRefreshMetadata(int timeoutMs) throws InterruptedException, ExecutionException, TimeoutException {
+    // We use both AdminClient timeout options and `future.get(timeout)`
+    // The former ensures the timeouts are enforced in the brokers
+    int remainingMs = timeoutMs;
+    long startMs = _time.milliseconds();
+
+    Set<String> topicNames = _adminClient.listTopics(new ListTopicsOptions().timeoutMs(remainingMs).listInternal(true))
+        .names().get(remainingMs, TimeUnit.MILLISECONDS);
+    long endMs = _time.milliseconds();
+    long elapsedTimeMs = endMs - startMs;
+    remainingMs -= elapsedTimeMs;
+
+    DescribeClusterResult result = _adminClient.describeCluster(new DescribeClusterOptions().timeoutMs(remainingMs));
+    Collection<Node> nodes = result.nodes().get(remainingMs, TimeUnit.MILLISECONDS);
+    String clusterId = result.clusterId().get(remainingMs, TimeUnit.MILLISECONDS);
+    Node controller = result.controller().get(remainingMs, TimeUnit.MILLISECONDS);
+    elapsedTimeMs = _time.milliseconds() - endMs;
+    remainingMs -= elapsedTimeMs;
+
+    Map<String, TopicDescription> topicDescriptions = _adminClient.describeTopics(topicNames, new DescribeTopicsOptions().timeoutMs(remainingMs))
+        .all().get(remainingMs, TimeUnit.MILLISECONDS);
+
+    return cluster(clusterId, nodes, controller, topicDescriptions);
+  }
+
+  private Cluster cluster(String clusterId, Collection<Node> nodes, Node controller,
+                          Map<String, TopicDescription> describeTopicResult) {
+    List<PartitionInfo> partitionInfos = new LinkedList<>();
+    Set<String> internalTopics = new HashSet<>();
+
+    for (Map.Entry<String, TopicDescription> topicDescription : describeTopicResult.entrySet()) {
+      if (topicDescription.getValue().isInternal()) {
+        internalTopics.add(topicDescription.getKey());
+      }
+
+      for (TopicPartitionInfo topicPartitionInfo : topicDescription.getValue().partitions()) {
+        PartitionInfo partitionInfo = new PartitionInfo(topicDescription.getKey(),
+            topicPartitionInfo.partition(),
+            topicPartitionInfo.leader(),
+            topicPartitionInfo.replicas().toArray(new Node[0]),
+            topicPartitionInfo.isr().toArray(new Node[0]),
+            topicPartitionInfo.replicas()
```

Comment:
`ISR` indicates the subset of the replicas that are in sync, that is caught-up to the leader and ready to take over as leader if the leader should fail. Hence, if a replica is not in sync, it could be either (1) offline, or (2) online, but falling behind. So IIUC, this filter not necessarily yields the offline replicas. Am I missing something?